src/components/Microsoft.DotNet.UpgradeAssistant.MSBuild/PackageLoader.cs

```diff
@@ -68,68 +79,35 @@ public async Task<bool> DoesPackageSupportTargetFrameworksAsync(NuGetReference p
+        public Task<IEnumerable<NuGetReference>> GetNewerVersionsAsync(NuGetReference reference, IEnumerable<TargetFrameworkMoniker> tfms, bool latestMinorAndBuildOnly, CancellationToken token)
         {
             if (reference is null)
             {
                 throw new ArgumentNullException(nameof(reference));
             }
 
-            var versions = new List<NuGetVersion>();
-
-            // Query each package source for listed versions of the given package name
-            foreach (var source in _packageSources)
-            {
-                try
-                {
-                    var metadata = await GetSourceRepository(source).GetResourceAsync<PackageMetadataResource>(token).ConfigureAwait(false);
-                    var searchResults = await CallWithRetryAsync(() => metadata.GetMetadataAsync(reference.Name, includePrerelease: true, includeUnlisted: false, _cache, _nugetLogger, token)).ConfigureAwait(false);
-                    versions.AddRange(searchResults.Select(r => r.Identity.Version));
-                }
-                catch (NuGetProtocolException)
-                {
-                    _logger.LogWarning("Failed to get package versions from source {PackageSource} due to a NuGet protocol error", source.Source);
-                }
-                catch (HttpRequestException exc)
-                {
-                    _logger.LogWarning("Failed to get package versions from source {PackageSource} due to an HTTP error ({StatusCode})", source.Source, exc.StatusCode);
-                }
-            }
-
-            // Filter to only include versions higher than the user's current version and,
-            // optionally, only the highest minor/build for each major version
-            var currentVersion = reference.GetNuGetVersion();
-            var filteredVersions = versions.Distinct().Where(v => v > currentVersion);
-            var versionsToReturn = latestMinorAndBuildOnly
-                ? filteredVersions.GroupBy(v => v.Major).Select(v => v.Where(v => !v.IsPrerelease).Max() ?? v.Max()!)
-                : filteredVersions;
+            return SearchByNameAsync(reference.Name, tfms, currentVersion: reference.GetNuGetVersion(), latestMinorAndBuildOnly: latestMinorAndBuildOnly, token: token);
+        }
 
-            _logger.LogDebug("Found versions for package {PackageName}: {PackageVersions}", reference.Name, versionsToReturn);
+        public async Task<NuGetReference?> GetLatestVersionAsync(string packageName, IEnumerable<TargetFrameworkMoniker> tfms, bool includePreRelease, CancellationToken token)
+        {
+            var result = await SearchByNameAsync(packageName, tfms, includePreRelease, token: token).ConfigureAwait(false);
 
-            return versionsToReturn.OrderBy(v => v).Select(v => reference with { Version = v.ToNormalizedString() });
+            return result.FirstOrDefault();
```

Comment:
Should this be `LastOrDefault`? It looks like `SearchByNameAsync` orders results least to greatest.